drivers/arm-cca/ukrsi/rsi.c

```diff
@@ -0,0 +1,288 @@
+				   unsigned long *new_base)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_IPA_STATE_SET;
+	args.a1 = base;
+	args.a2 = (top - base);
+	args.a3 = ripas;
+	args.a4 = flags;
+
+	smccc_invoke(&args);
+
+	*new_base = args.a1;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_measurement_extend(unsigned long index, unsigned long size,
+					unsigned long value[8])
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_MEASUREMENT_EXTEND;
+	args.a1 = index;
+	args.a2 = size;
+	args.a3 = value[0];
+	args.a4 = value[1];
+	args.a5 = value[2];
+	args.a6 = value[3];
+	args.a7 = value[4];
+	args.a8 = value[5];
+	args.a9 = value[6];
+	args.a10 = value[7];
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_measurement_read(unsigned long index,
+				      unsigned long value[8])
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_MEASUREMENT_READ;
+	args.a1 = index;
+
+	smccc_invoke(&args);
+
+	value[0] = args.a1;
+	value[1] = args.a2;
+	value[2] = args.a3;
+	value[3] = args.a4;
+	value[4] = args.a5;
+	value[5] = args.a6;
+	value[6] = args.a7;
+	value[7] = args.a8;
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_realm_config(unsigned long config)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_REALM_CONFIG;
+	args.a1 = config;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+unsigned long uk_rsi_version(void)
+{
+	struct smccc_args args = {0};
+
+	args.a0 = RSI_VERSION;
+
+	smccc_invoke(&args);
+
+	return args.a0;
+}
+
+void uk_rsi_init(struct ukplat_bootinfo *bi)
+{
+	struct rsi_realm_config __attribute__((aligned(PAGE_SIZE))) config;
+	const struct ukplat_memregion_desc *mrd;
+	__paddr_t pstart, pend, len;
+	unsigned long version;
+	unsigned long ret;
+	__u32 i;
+
+	version = uk_rsi_version();
+	uk_pr_info("RSI version %lu.%lu\n", (version >> 16),
+		   (version & 0xFFFF));
+
+	for (i = 0; i < bi->mrds.count; i++) {
+		mrd = &bi->mrds.mrds[i];
+
+		pstart = ALIGN_UP(mrd->pbase, __PAGE_SIZE);
+		len = ALIGN_UP(mrd->len, __PAGE_SIZE);
+		pend   = pstart + len;
+
+		ret = uk_rsi_setup_memory(pstart, pend, RSI_RAM);
+		if (ret != RSI_SUCCESS) {
+			uk_pr_err("RSI setup memory failed\n");
+			return;
+		}
+	}
+
+	ret = uk_rsi_realm_config((unsigned long)&config);
+	if (ret != RSI_SUCCESS) {
+		uk_pr_err("RSI realm config failed\n");
+		return;
+	}
+
+	uk_pr_info("RSI realm config ipa_width %lu, hash_algo %lu:\n",
+		   config.ipa_width, config.hash_algo);
+
+	/* set the mask of the unprotected bit */
+	unprotected_mask = (1UL) << (config.ipa_width - 1);
+
+	uk_pr_info("RSI initialized unprotected_mask %lx\n", unprotected_mask);
+}
+
+unsigned long
+uk_rsi_setup_memory(unsigned long base, unsigned long end, unsigned char ripas)
+{
+	unsigned long ret;
+	unsigned long new_base;
+
+	/* Iterate over the memory space to set RIPAS */
+	while (base != end) {
+		ret = uk_rsi_ipa_state_set(base, end, ripas, 0, &new_base);
+		base = new_base;
+		if (ret != RSI_SUCCESS)
+			break;
+	}
+	return ret;
+}
+
+void uk_rsi_setup_device(void)
```

Comment:
This function makes a very strong assumption about the platform, i.e. that the device memory is 1x page at the beginning of the RAM. To remove this dependency you can update the prototype to:
```
int uk_rsi_setup_device(__u64 base, __sz size)
```
and also add assertions that the base and size are page-aligned.

Then you can define the device regions in `plat/kvm/include/kvm-arm64/image.h`:
```
diff --git a/plat/kvm/include/kvm-arm64/image.h b/plat/kvm/include/kvm-arm64/image.h
index 8464601b..99f7b17a 100644
--- a/plat/kvm/include/kvm-arm64/image.h
+++ b/plat/kvm/include/kvm-arm64/image.h
@@ -7,11 +7,15 @@
 #define __KVM_ARM64_IMAGE_H__
 
 #if CONFIG_KVM_VMM_QEMU
-#define RAM_BASE_ADDR 0x40000000
+#define RAM_BASE_ADDR          0x40000000
+#define DEVICE_BASE_ADDR       0x08000000
+#define DEVICE_LENGTH          0x38000000
 #elif CONFIG_KVM_VMM_FIRECRACKER
-#define RAM_BASE_ADDR 0x80000000
+#define RAM_BASE_ADDR          0x80000000
+#define DEVICE_BASE_ADDR       0x00000000
+#define DEVICE_LENGTH          0x80000000
 #endif /* CONFIG_KVM_VMM_FIRECRACKER  */
 
-#define DTB_RESERVED_SIZE 0x100000
+#define DTB_RESERVED_SIZE      0x100000
 
 #endif /*  __KVM_ARM64_IMAGE_H__ */
```
And finally update `plat/common/memory.h` as:
```
diff --git a/plat/common/memory.c b/plat/common/memory.c
index 0960cdc7..384bb31f 100644
--- a/plat/common/memory.c
+++ b/plat/common/memory.c
@@ -33,6 +33,7 @@
  */
 #include <stdbool.h>
 #include <stddef.h>
+#include <kvm-arm64/image.h>
 #include <uk/plat/common/sections.h>
 #include <uk/plat/common/bootinfo.h>
 #include <uk/asm/limits.h>
@@ -425,7 +426,7 @@ int ukplat_mem_init(void)
        ukplat_memregion_list_delete(&bi->mrds, 0);
 
 #if CONFIG_LIBUKRSI
-       uk_rsi_setup_device();
+       uk_rsi_setup_device(DEVICE_BASE_ADDR, DEVICE_LENGTH);
 #endif /* CONFIG_LIBUKRSI */
 
        return 0;
```

**Important:** Is marking all device memory as shared a hard requirement of CCA (ie so that the device can access that memory), or is it only required for devices emulated by the host? How do devices fit into the security model of the CCA? Is it guaranteed that device memory is always visible by the host?

Either way, I expect that this function will be deprecated later when we introduce `ioremap`, as we will be able to set all device attributes (inc. `PAGE_NS_SHARED`) from there.